Tweak Tracks docs generation,

tomalec opened the pull request:
<!-- You can erase any parts of this template not applicable to your pull request. -->

<!-- Reference issue(s) that this PR fixes or addresses -->
related to #232 

### Changes proposed in this pull request
<!-- Explain what this PR adds or changes, why, and how this will impact users. -->
- sort events by name
- add source file lineno to the link
- add a line break after the end disclaimer
- support empty descriptions
- document links of Track event emitters
- added a plugin, to handle `@fires` descriptions better - support `.:,#\n`

#### Screenshots
<!--- Optional --->

### Detailed test instructions
<!-- Add steps to confirm the fix or change. -->
0. Do some changes in `@events` descriptions
1. `npm run doc:tracking`
2. Review the `TRACKING.md`

<!-- Please add details of other areas that might be impacted by the change. -->

### Changelog note
<!-- Changelog notes highlight what's fixed or added in a release. -->

tomalec committed:
Tweak Tracks docs generation,

- sort events by name
- add sourcefile lineno to the link
- add line break after the end disclaimer
- support empty descriptions

tomalec committed:
Document lins of Track event emitters.

eason9487 commented on woo-tracking-jsdoc/publish.js:
```diff
@@ -40,34 +58,64 @@ exports.publish = function ( data ) {
+				} );
+			}
+
+			// Find all palces that fires the event.
```
💅 
Seems a typo?
```suggestion
			// Find all places that fires the event.
```

eason9487 commented on woo-tracking-jsdoc/publish.js:
```diff
@@ -40,34 +58,64 @@ exports.publish = function ( data ) {
+			const emitters = new Map();
+			// TaffyDB#has is buggy https://github.com/typicaljoe/taffydb/issues/19, so let's filter it manually.
+			data( { fires: { isArray: true } } ).each( ( emitter ) => {
+				const firesCurrent = emitter.fires.filter( ( fires ) =>
+					fires.startsWith( 'event:' + symbol.name )
+				)
+				if ( firesCurrent.length ) {
+					emitters.set(
+						emitter,
+						// Consider everything after the event name [and whitespace] an additional description.
+						// Sanitize the descriptions, trim empty ones.
+						firesCurrent.map( description => description.match( /^(\S+\s*)(.*)/ )[ 2 ])
+					);
+				}
```
❓ 
I'm trying to understand what formatting would appear as a multi-line description, but it seems that `#` and `.` symbols for `@fires` are part of its format. If the partial formatting is borrowed as a description, there may be more than one of those symbols in the description, and then resulting in unintended parsing results. For examples:
- Multiple sentences in a single line. Or, if there is anything after the first `.`, it will have the same result
    ```
    * @fires wcadmin_pfw_get_started_faq whenever the FAQ is toggled. The second sentence.
    ```
- Multi-line descriptions
    ```
    * @fires wcadmin_pfw_get_started_faq whenever the FAQ is toggled.
        The second sentence.
    ```
- Using `#`. (Looks like it follows the same parsing rule of `.`)
    ```
    * @fires wcadmin_pfw_get_started_faq whenever the #FAQ is toggled
    ```

These usages will cause different formats of strings `emitter.fires` and mismatch the `filter()` in line 91.

tomalec replied:
Sorry, I should've mentioned that.
- To use `.` and `:` you need to use the full name path for the event, like:
   ```
   * @fires event:wcadmin_pfw_get_started_faq whenever the FAQ is toggled. The second sentence `{ foo: bar, baz: fiz}`.
   ```
   - Currently, this breaks `lint:js` as we use an outdated version of `eslint-plugin-jsdoc` without this fix https://github.com/gajus/eslint-plugin-jsdoc/issues/750. I was thinking of installing the updated version of `eslint-plugin-jsdoc` once we have the use case for that, as I hope `woocommerce/eslint…` may catch up in the meantime.
- I missed multiline support. I can add it by updating the regexp to `description.match( /^(\S+\s*)(.*)/ )[ 2 ]` - but it again has the `event:` problem  with dots in following lines. - I guess it's just a thing in `jsdoc` parser, as it really does not match catch the description, just considers everything a name :(	I may try to cover it by PR to jsdoc or separate JSDoc plugin, but that would definitely be a separate PR :scroll: 
- `#` and `,` are even more problematic, as I have found no easy workaround.

I think in this PR I may just fix the multiline support. Then, work on better parsing of `#,.:` in fires descriptions, and actually parsing first desriptions in a separate PR.

tomalec replied:
I created a new plugin in https://github.com/woocommerce/pinterest-for-woocommerce/pull/254/commits/d42f3c52e4e1c34213da143acb5ae4519eabfcb1

Thanks to it, 
 - there is no need to use the full event name (which was incompatible with our bugged version of eslint)
 - we can use `.;#` in descriptions
 - we can use line breaks
 - there is a strict distinction between name and desc.


I think as we reached 2 plugins + template, it's a fair time to think about extracting this tool to a separate repo.

I was thinking of making a PR to jsdoc/jsdoc with the `@fires` fix/feature. But I think, that may be too much effort, given there is a jsdoc 4.x line coming, and I already spend much time on jsdoc instead of actual events.

eason9487 reviewed (commented):
💅 
This file is not included in the check scope of JS lint. But I think it could still be adjusted the format to avoid distracted editor errors?

![image](https://user-images.githubusercontent.com/17420811/141766304-2b0e76e6-5ebf-4e1a-b8bc-55bb935825f3.png)

tomalec committed:
Fix a typo in code comment woo-tracking-jsdoc/publish.js

Co-authored-by: Eason <eason.su.tw@gmail.com>

tomalec committed:
Fix whitespaces in `woo-tracking-jsdoc/publish.js`.

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/254#pullrequestreview-805853439

tomalec committed:
Match multiline `@fires` descriptions.

Addresses a bit of https://github.com/woocommerce/pinterest-for-woocommerce/pull/254#discussion_r749191354

tomalec commented:
@eason9487 Thanks for the review, would you mind another round?

tomalec committed:
Create JSDoc plugin to support `fires`' description

Address https://github.com/woocommerce/pinterest-for-woocommerce/pull/254#pullrequestreview-805853439

eason9487 reviewed (approved):
Tested locally and works well. LGTM.
